File: internal/projector/projector.go
package projector

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/junkawasaki/actordb-dokigoto/internal/eventstore"
	"github.com/junkawasaki/actordb-dokigoto/pkg/config"
)

// ProjectionDefinition defines a projection's schema and IVM rules
// Merkle DAG: sha256:projection_def_v1 - Declarative projection DSL
type ProjectionDefinition struct {
	Name            string                 `yaml:"name"`
	Sources         []SourceDefinition     `yaml:"sources"`
	StateSchema     map[string]interface{} `yaml:"state"`
	IVM             IVMConfig              `yaml:"ivm"`
	Security        SecurityConfig         `yaml:"security"`
	Materialization MaterializationConfig  `yaml:"materialization"`
}

// SourceDefinition defines an event source for projection
type SourceDefinition struct {
	Stream string `yaml:"stream"`
	Key    string `yaml:"key"`
	Filter string `yaml:"filter,omitempty"`
}

// IVMConfig configures Incremental View Maintenance
type IVMConfig struct {
	LateWindowMs   int64       `yaml:"late_window_ms"`
	WatermarkLagMs int64       `yaml:"watermark_lag_ms"`
	Delta          []DeltaRule `yaml:"delta"`
}

// DeltaRule defines how to update projection state from events
type DeltaRule struct {
	On     string `yaml:"on"`     // event type
	Update string `yaml:"update"` // update expression
}

// SecurityConfig defines security policies for the projection
type SecurityConfig struct {
	RLS      string   `yaml:"rls"`  // Row Level Security expression
	Mask     []string `yaml:"mask"` // Columns to mask
	TenantID string   `yaml:"tenant_id"`
}

// MaterializationConfig controls auto-promotion/demotion
type MaterializationConfig struct {
	PromoteIfQPS float64 `yaml:"promote_if_qps"`
	DemoteIfQPS  float64 `yaml:"demote_if_qps"`
}

// ProjectionState holds the current state of a projection
type ProjectionState struct {
	Name           string
	IsMaterialized bool
	LastUpdate     time.Time
	QPS            float64
	State          map[string]interface{}
	StateMu        sync.RWMutex
}

// ProjectionResult contains query results
type ProjectionResult struct {
	Data      interface{}
	Source    string // "materialized" or "ondemand"
	Latency   time.Duration
	Timestamp time.Time
}

// ProjectionEngine handles IVM with auto-materialization
// Process Network Node: projection_engine
// Dependencies: [event_stream, catalog_service]
// Outputs: [materialized_views, ondemand_results]
// SLO: p99_latency_200ms_ondemand_50ms_materialized
type ProjectionEngine struct {
	config        config.ProjectionConfig
	eventstore    *eventstore.EventStore
	projections   map[string]*ProjectionState
	projectionsMu sync.RWMutex
	definitions   map[string]*ProjectionDefinition
	definitionsMu sync.RWMutex
	workers       []*Worker
	eventChannel  chan eventstore.Event // Channel to receive events from EventStore
	running       bool
	ctx           context.Context
	cancel        context.CancelFunc
}

// Worker handles projection updates
type Worker struct {
	id      int
	engine  *ProjectionEngine
	queue   chan ProjectionUpdate
	running bool
}

// ProjectionUpdate represents an update to process
type ProjectionUpdate struct {
	ProjectionName string
	Event          eventstore.Event
	Priority       int // 0=interactive, 1=batch
}

// New creates a new ProjectionEngine
func New(cfg config.ProjectionConfig, es *eventstore.EventStore) (*ProjectionEngine, error) {
	ctx, cancel := context.WithCancel(context.Background())

	pe := &ProjectionEngine{
		config:      cfg,
		eventstore:  es,
		projections: make(map[string]*ProjectionState),
		definitions: make(map[string]*ProjectionDefinition),
		workers:     make([]*Worker, cfg.WorkerCount),
		running:     false,
		ctx:         ctx,
		cancel:      cancel,
	}

	// Initialize workers
	for i := 0; i < cfg.WorkerCount; i++ {
		pe.workers[i] = &Worker{
			id:     i,
			engine: pe,
			queue:  make(chan ProjectionUpdate, 1000),
		}
	}

	return pe, nil
}

// Start begins the ProjectionEngine operation
func (pe *ProjectionEngine) Start(ctx context.Context) error {
	pe.running = true

	// Subscribe to the event store to receive real-time events.
	pe.eventChannel = pe.eventstore.Subscribe()

	// Start workers
	for _, worker := range pe.workers {
		go worker.run()
	}

	// Start event subscription
	go pe.subscribeToEvents()

	log.Printf("ProjectionEngine started with %d workers", len(pe.workers))
	return nil
}

// Stop shuts down the ProjectionEngine
func (pe *ProjectionEngine) Stop() {
	pe.running = false
	pe.cancel()

	// Stop workers
	for _, worker := range pe.workers {
		if worker.running {
			close(worker.queue)
		}
	}

	// Unsubscribe from the event store to release resources.
	if pe.eventChannel != nil {
		pe.eventstore.Unsubscribe(pe.eventChannel)
	}

	log.Println("ProjectionEngine stopped")
}

// RegisterProjection registers a new projection definition
func (pe *ProjectionEngine) RegisterProjection(def *ProjectionDefinition) error {
	pe.definitionsMu.Lock()
	pe.definitions[def.Name] = def
	pe.definitionsMu.Unlock()

	pe.projectionsMu.Lock()
	pe.projections[def.Name] = &ProjectionState{
		Name:           def.Name,
		IsMaterialized: false,
		LastUpdate:     time.Now(),
		QPS:            0,
		State:          make(map[string]interface{}),
	}
	pe.projectionsMu.Unlock()

	log.Printf("Registered projection: %s", def.Name)
	return nil
}

// Query executes a projection query
func (pe *ProjectionEngine) Query(ctx context.Context, projectionName string, params map[string]interface{}) (*ProjectionResult, error) {
	if !pe.running {
		return nil, fmt.Errorf("projection engine not running")
	}

	start := time.Now()

	pe.projectionsMu.RLock()
	proj, exists := pe.projections[projectionName]
	pe.projectionsMu.RUnlock()

	if !exists {
		return nil, fmt.Errorf("projection %s not found", projectionName)
	}

	proj.StateMu.RLock()
	state := make(map[string]interface{})
	for k, v := range proj.State {
		state[k] = v
	}
	proj.StateMu.RUnlock()

	result := &ProjectionResult{
		Data:      state,
		Timestamp: time.Now(),
		Latency:   time.Since(start),
	}

	if proj.IsMaterialized {
		result.Source = "materialized"
	} else {
		result.Source = "ondemand"
		// Update QPS for auto-promotion logic
		pe.updateQPS(projectionName)
	}

	return result, nil
}

// subscribeToEvents subscribes to eventstore events and routes them to workers
func (pe *ProjectionEngine) subscribeToEvents() {
	log.Println("ProjectionEngine is now listening for events from EventStore...")
	for {
		select {
		case <-pe.ctx.Done():
			log.Println("Stopping event subscription.")
			return
		case event, ok := <-pe.eventChannel:
			if !ok {
				log.Println("Event channel closed, stopping subscription.")
				return
			}
			pe.routeEvent(event)
		}
	}
}

// routeEvent routes an event to appropriate projections
func (pe *ProjectionEngine) routeEvent(event eventstore.Event) {
	if !pe.running {
		return
	}

	pe.definitionsMu.RLock()
	for name, def := range pe.definitions {
		if pe.shouldProcessEvent(def, event) {
			update := ProjectionUpdate{
				ProjectionName: name,
				Event:          event,
				Priority:       1, // Default to batch priority
			}

			// Route to worker (simple round-robin for MVP)
			workerIndex := len(name) % len(pe.workers)
			select {
			case pe.workers[workerIndex].queue <- update:
			default:
				log.Printf("Worker queue full, dropping update for %s", name)
			}
		}
	}
	pe.definitionsMu.RUnlock()
}

// shouldProcessEvent checks if a projection should process an event
func (pe *ProjectionEngine) shouldProcessEvent(def *ProjectionDefinition, event eventstore.Event) bool {
	for _, source := range def.Sources {
		if source.Stream == event.EventType {
			return true
		}
	}
	return false
}

// updateQPS updates QPS tracking for auto-promotion
func (pe *ProjectionEngine) updateQPS(projectionName string) {
	pe.projectionsMu.Lock()
	defer pe.projectionsMu.Unlock()

	if proj, exists := pe.projections[projectionName]; exists {
		proj.QPS += 1.0 // Simplified QPS tracking
	}
}

// run executes projection updates (Worker method)
func (w *Worker) run() {
	w.running = true
	log.Printf("Worker %d started", w.id)

	for {
		select {
		case update, ok := <-w.queue:
			if !ok {
				w.running = false
				return
			}
			w.processUpdate(update)
		case <-w.engine.ctx.Done():
			w.running = false
			return
		}
	}
}

// processUpdate applies a projection update
func (w *Worker) processUpdate(update ProjectionUpdate) {
	// Get projection definition
	w.engine.definitionsMu.RLock()
	_, exists := w.engine.definitions[update.ProjectionName]
	w.engine.definitionsMu.RUnlock()

	if !exists {
		log.Printf("Definition not found for projection: %s", update.ProjectionName)
		return
	}

	// Apply IVM rules
	w.engine.projectionsMu.Lock()
	proj := w.engine.projections[update.ProjectionName]
	w.engine.projectionsMu.Unlock()

	proj.StateMu.Lock()
	// MVP: Simple state updates based on event type
	switch update.Event.EventType {
	case "order_created":
		if count, ok := proj.State["orders"].(int); ok {
			proj.State["orders"] = count + 1
		} else {
			proj.State["orders"] = 1
		}
	case "order_cancelled":
		if count, ok := proj.State["orders"].(int); ok && count > 0 {
			proj.State["orders"] = count - 1
		}
	}
	proj.LastUpdate = time.Now()
	proj.StateMu.Unlock()

	// Check for auto-promotion
	w.checkAutoPromotion(update.ProjectionName, proj)
}

// checkAutoPromotion checks if a projection should be promoted to materialized
func (w *Worker) checkAutoPromotion(name string, proj *ProjectionState) {
	if !proj.IsMaterialized && proj.QPS >= w.engine.config.AutoPromoteQPSThreshold {
		log.Printf("Auto-promoting projection %s (QPS: %.2f)", name, proj.QPS)
		proj.IsMaterialized = true
		// Reset QPS after promotion
		proj.QPS = 0
	}
}

File: demo/nextjs-supabase/src/lib/actordb/projector/projector.ts
import { ProjectionConfig } from '../config/config';
import { EventStore, Event } from '../eventstore';

// ProjectionDefinition defines a projection's schema and IVM rules
// Merkle DAG: sha256:projection_def_v1 - Declarative projection DSL
export interface ProjectionDefinition {
  name: string;
  sources: SourceDefinition[];
  stateSchema?: { [key: string]: any };
  ivm: IVMConfig;
  security: SecurityConfig;
  materialization: MaterializationConfig;
}

// SourceDefinition defines an event source for projection
export interface SourceDefinition {
  stream: string;
  key?: string;
  filter?: string;
}

// IVMConfig configures Incremental View Maintenance
export interface IVMConfig {
  lateWindowMs: number;
  watermarkLagMs: number;
  delta: DeltaRule[];
}

// DeltaRule defines how to update projection state from events
export interface DeltaRule {
  on: string; // event type
  update: string; // update expression
}

// SecurityConfig defines security policies for the projection
export interface SecurityConfig {
  rls: string; // Row Level Security expression
  mask: string[]; // Columns to mask
  tenantId?: string;
}

// MaterializationConfig controls auto-promotion/demotion
export interface MaterializationConfig {
  promoteIfQPS: number;
  demoteIfQPS: number;
}

// ProjectionState holds the current state of a projection
export interface ProjectionState {
  name: string;
  isMaterialized: boolean;
  lastUpdate: Date;
  qps: number;
  state: { [key: string]: any };
}

// ProjectionResult contains query results
export interface ProjectionResult {
  data: any;
  source: 'materialized' | 'ondemand';
  latency: number; // in milliseconds
  timestamp: Date;
}

// Worker handles projection updates
class Worker {
  private id: number;
  private engine: ProjectionEngine;
  private queue: ProjectionUpdate[] = [];
  private running: boolean = false;
  private processing: boolean = false;

  constructor(id: number, engine: ProjectionEngine) {
    this.id = id;
    this.engine = engine;
  }

  start(): void {
    this.running = true;
    this.process();
  }

  stop(): void {
    this.running = false;
  }

  enqueue(update: ProjectionUpdate): void {
    this.queue.push(update);
    this.process();
  }

  private async process(): Promise<void> {
    if (this.processing || !this.running) {
      return;
    }

    this.processing = true;

    while (this.running && this.queue.length > 0) {
      const update = this.queue.shift();
      if (update) {
        try {
          await this.processUpdate(update);
        } catch (err) {
          console.error(`Worker ${this.id} error processing update: ${err}`);
        }
      }
    }

    this.processing = false;
  }

  private async processUpdate(update: ProjectionUpdate): Promise<void> {
    // Get projection definition
    const def = this.engine.getProjectionDefinition(update.projectionName);
    if (!def) {
      console.error(`Definition not found for projection: ${update.projectionName}`);
      return;
    }

    // Apply IVM rules
    const proj = this.engine.getProjectionState(update.projectionName);
    if (!proj) {
      return;
    }

    // MVP: Simple state updates based on event type
    switch (update.event.eventType) {
      case 'order_created':
        if (typeof proj.state.orders === 'number') {
          proj.state.orders += 1;
        } else {
          proj.state.orders = 1;
        }
        break;
      case 'order_cancelled':
        if (typeof proj.state.orders === 'number' && proj.state.orders > 0) {
          proj.state.orders -= 1;
        }
        break;
    }

    proj.lastUpdate = new Date();

    // Check for auto-promotion
    this.checkAutoPromotion(update.projectionName, proj);
  }

  private checkAutoPromotion(name: string, proj: ProjectionState): void {
    if (!proj.isMaterialized && proj.qps >= this.engine.getConfig().auto_promote_qps_threshold) {
      console.log(`Auto-promoting projection ${name} (QPS: ${proj.qps})`);
      proj.isMaterialized = true;
      // Reset QPS after promotion
      proj.qps = 0;
    }
  }
}

// ProjectionUpdate represents an update to process
interface ProjectionUpdate {
  projectionName: string;
  event: Event;
  priority: number; // 0=interactive, 1=batch
}

// ProjectionEngine handles IVM with auto-materialization
// Process Network Node: projection_engine
// Dependencies: [event_stream, catalog_service]
// Outputs: [materialized_views, ondemand_results]
// SLO: p99_latency_200ms_ondemand_50ms_materialized
export class ProjectionEngine {
  private config: ProjectionConfig;
  private eventstore: EventStore;
  private projections: Map<string, ProjectionState> = new Map();
  private definitions: Map<string, ProjectionDefinition> = new Map();
  private workers: Worker[] = [];
  private running: boolean = false;
  private unsubscribe?: () => void;

  constructor(config: ProjectionConfig, eventstore: EventStore) {
    this.config = config;
    this.eventstore = eventstore;

    // Initialize workers
    for (let i = 0; i < config.worker_count; i++) {
      this.workers.push(new Worker(i, this));
    }
  }

  // Start begins the ProjectionEngine operation
  async start(): Promise<void> {
    this.running = true;

    // Start workers
    for (const worker of this.workers) {
      worker.start();
    }

    // Subscribe to the event store
    this.unsubscribe = this.eventstore.subscribe((event) => {
      this.routeEvent(event);
    });

    console.log(`ProjectionEngine started with ${this.workers.length} workers`);
  }

  // Stop shuts down the ProjectionEngine
  async stop(): Promise<void> {
    this.running = false;

    // Stop workers
    for (const worker of this.workers) {
      worker.stop();
    }

    // Unsubscribe from the event store
    if (this.unsubscribe) {
      this.unsubscribe();
    }

    console.log('ProjectionEngine stopped');
  }

  // RegisterProjection registers a new projection definition
  registerProjection(def: ProjectionDefinition): void {
    this.definitions.set(def.name, def);

    this.projections.set(def.name, {
      name: def.name,
      isMaterialized: false,
      lastUpdate: new Date(),
      qps: 0,
      state: {},
    });

    console.log(`Registered projection: ${def.name}`);
  }

  // Query executes a projection query
  // Merkle DAG: sha256:projection_query_v1 - Materialized/ondemand projection query execution
  async query(projectionName: string, params?: { [key: string]: any }): Promise<ProjectionResult> {
    if (!this.running) {
      throw new Error('projection engine not running');
    }

    const start = Date.now();

    const proj = this.projections.get(projectionName);
    if (!proj) {
      throw new Error(`projection ${projectionName} not found`);
    }

    const result: ProjectionResult = {
      data: { ...proj.state },
      timestamp: new Date(),
      latency: Date.now() - start,
      source: proj.isMaterialized ? 'materialized' : 'ondemand',
    };

    if (!proj.isMaterialized) {
      // Update QPS for auto-promotion logic
      this.updateQPS(projectionName);
    }

    return result;
  }

  // Internal methods for workers
  getConfig(): ProjectionConfig {
    return this.config;
  }

  getProjectionDefinition(name: string): ProjectionDefinition | undefined {
    return this.definitions.get(name);
  }

  getProjectionState(name: string): ProjectionState | undefined {
    return this.projections.get(name);
  }

  private routeEvent(event: Event): void {
    if (!this.running) {
      return;
    }

    for (const [name, def] of this.definitions) {
      if (this.shouldProcessEvent(def, event)) {
        const update: ProjectionUpdate = {
          projectionName: name,
          event,
          priority: 1, // Default to batch priority
        };

        // Route to worker (simple round-robin for MVP)
        const workerIndex = name.length % this.workers.length;
        this.workers[workerIndex].enqueue(update);
      }
    }
  }

  private shouldProcessEvent(def: ProjectionDefinition, event: Event): boolean {
    for (const source of def.sources) {
      if (source.stream === event.eventType) {
        return true;
      }
    }
    return false;
  }

  private updateQPS(projectionName: string): void {
    const proj = this.projections.get(projectionName);
    if (proj) {
      proj.qps += 1.0; // Simplified QPS tracking
    }
  }
}
